tools/copyblocks/bucket.go

```diff
@@ -0,0 +1,232 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package main
+
+import (
+	"context"
+	"io"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/container"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/sas"
+	"github.com/grafana/dskit/backoff"
+	"github.com/pkg/errors"
+	"google.golang.org/api/iterator"
+)
+
+type bucket interface {
+	Get(ctx context.Context, name string) (io.ReadCloser, error)
+	Copy(ctx context.Context, name string, dstBucket bucket) error
+	ListPrefix(ctx context.Context, prefix string, recursive bool) ([]string, error)
+	UploadMarkerFile(ctx context.Context, name string) error
+	Name() string
+}
+
+type gcsBucket struct {
+	storage.BucketHandle
+	name string
+}
+
+func (bkt *gcsBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	obj := bkt.Object(name)
+	r, err := obj.NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func (bkt *gcsBucket) Copy(ctx context.Context, name string, dstBucket bucket) error {
+	d, ok := dstBucket.(*gcsBucket)
+	if !ok {
+		return errors.New("destination bucket wasn't a gcs bucket")
+	}
+	srcObj := bkt.Object(name)
+	dstObject := d.BucketHandle.Object(name)
+	copier := dstObject.CopierFrom(srcObj)
+	_, err := copier.Run(ctx)
+	return err
+}
+
+func (bkt *gcsBucket) ListPrefix(ctx context.Context, prefix string, recursive bool) ([]string, error) {
+	if len(prefix) > 0 && prefix[len(prefix)-1:] != delim {
+		prefix = prefix + delim
+	}
+
+	q := &storage.Query{
+		Prefix: prefix,
+	}
+	if !recursive {
+		q.Delimiter = delim
+	}
+
+	var result []string
+
+	it := bkt.Objects(ctx, q)
+	for {
+		obj, err := it.Next()
+
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "listPrefix: error listing %v", prefix)
+		}
+
+		path := ""
+		if obj.Prefix != "" { // synthetic directory, only returned when recursive=false
+			path = obj.Prefix
+		} else {
+			path = obj.Name
+		}
+
+		if strings.HasPrefix(path, prefix) {
+			path = strings.TrimPrefix(path, prefix)
+		} else {
+			return nil, errors.Errorf("listPrefix: path has invalid prefix: %v, expected prefix: %v", path, prefix)
+		}
```

Comment:
Does this ever happen? Don't we set the prefix in the `Query` parameters above?

Reply:
I can go either way on this one. Considering we're in a bit of a weird position reimplementing listing logic for a tool I'm not necessarily against it. I preserved the sanity check here because it was present in the GCS implementation and copying the wrong thing is probably the worst thing this tool could do.

Reply:
Makes sense, happy to keep the check then